core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/kopt/KOptDescriptor.java

```diff
@@ -0,0 +1,504 @@
+     * {@link KOptDescriptor#removedEdgeIndexToTourOrder}[inverseRemovedEdgeIndexToTourOrder[i]] == i
+     *
+     */
+    private final int[] inverseRemovedEdgeIndexToTourOrder;
+
+    /**
+     * Maps the index of a removed edge endpoint to its corresponding added edge other endpoint. For instance,
+     * if the removed edges are (a, b), (c, d), (e, f) and the added edges are (a, d), (c, f), (e, b), then
+     * <br />
+     * removedEdges = [null, a, b, c, d, e, f] <br />
+     * addedEdgeToOtherEndpoint = [null, 4, 5, 6, 1, 2, 3] <br />
+     * <br />
+     * For any valid removedEdges index, (removedEdges[index], removedEdges[addedEdgeToOtherEndpoint[index]])
+     * is an edge added by this K-Opt move.
+     */
+    private final int[] addedEdgeToOtherEndpoint;
+
+    private static int[] computeInEdgesForSequentialMove(Object[] removedEdges) {
+        int[] out = new int[removedEdges.length];
+        int k = (removedEdges.length - 1) >> 1;
+
+        out[1] = removedEdges.length - 1;
+        out[removedEdges.length - 1] = 1;
+        for (int i = 1; i < k; i++) {
+            out[2 * i + 1] = 2 * i;
+            out[2 * i] = 2 * i + 1;
+        }
+
+        return out;
+    }
+
+    /**
+     * Create a sequential {@link KOptDescriptor} from the given removed edges.
+     *
+     * @param removedEdges The edges removed from the tour. The added edges will
+     *        be formed from opposite endpoints for consecutive edges.
+     * @param endpointToSuccessorFunction A {@link Function} that maps an endpoint to its successor
+     * @param betweenPredicate A {@link TriPredicate} that return true if and only if its middle
+     *        argument is between its first and last argument when the tour is
+     *        taken in the successor direction.
+     */
+    public KOptDescriptor(
+            Object[] removedEdges,
+            Function<Object, Object> endpointToSuccessorFunction,
+            TriPredicate<Object, Object, Object> betweenPredicate) {
+        this(removedEdges, computeInEdgesForSequentialMove(removedEdges), endpointToSuccessorFunction, betweenPredicate);
+    }
+
+    /**
+     * Create as sequential or non-sequential {@link KOptDescriptor} from the given removed
+     * and added edges.
+     *
+     * @param removedEdges The edges removed from the tour.
+     * @param addedEdgeToOtherEndpoint The edges added to the tour.
+     * @param endpointToSuccessorFunction A {@link Function} that maps an endpoint to its successor
+     * @param betweenPredicate A {@link TriPredicate} that return true if and only if its middle
+     *        argument is between its first and last argument when the tour is
+     *        taken in the successor direction.
+     */
+    public KOptDescriptor(
+            Object[] removedEdges,
+            int[] addedEdgeToOtherEndpoint,
+            Function<Object, Object> endpointToSuccessorFunction,
+            TriPredicate<Object, Object, Object> betweenPredicate) {
+        int i, j;
```

Comment:
Let's increase variable locality by declaring these closer to where they are used.